### PR TITLE
build: Detect if char equals int8_t

### LINKED ISSUE #13576: Can't compile bitcoin for SmartOS - recipe for target 'consensus/libbitcoinconsensus_la-merkle.lo' failed
I am trying to compile the latest bitcoin source from GitHub on the latest SmartOS and do not know how to solve the following compile error...

```
/opt/local/src/bitcoin]# uname -a
SunOS a36dbe40-5628-e6ea-f7b8-8d89982fe9bf 5.11 joyent_20180525T172343Z i86pc i386 i86pc Solaris

/opt/local/src/bitcoin]# gcc -v
Using built-in specs.
COLLECT_GCC=/opt/local/gcc49/bin/gcc
COLLECT_LTO_WRAPPER=/opt/local/gcc49/libexec/gcc/x86_64-sun-solaris2.11/4.9.3/lto-wrapper
Target: x86_64-sun-solaris2.11
Configured with: ../gcc-4.9.3/configure --enable-languages='c obj-c++ objc go fortran c++' --enable-shared --enable-long-long --with-local-prefix=/opt/local --enable-libssp --enable-threads=posix --with-boot-ldflags='-static-libstdc++ -static-libgcc -Wl,-R/opt/local/lib ' --disable-nls --with-gxx-include-dir=/opt/local/gcc49/include/c++/ --without-gnu-ld --with-ld=/usr/bin/ld --with-gnu-as --with-as=/opt/local/bin/gas --prefix=/opt/local/gcc49 --build=x86_64-sun-solaris2.11 --host=x86_64-sun-solaris2.11 --infodir=/opt/local/gcc49/info --mandir=/opt/local/gcc49/man
Thread model: posix
gcc version 4.9.3 (GCC)

/opt/local/src/bitcoin]# ./configure --with-gui=no --disable-tests --enable-cxx --disable-shared --with-pic --prefix=/usr/local --with-incompatible-bdb --target=x86_64-sun-solaris2.11
...
Options used to compile and link:
  with wallet   = yes
  with gui / qt = no
  with zmq      = no
  with test     = no
  with bench    = yes
  with upnp     = auto
  use asm       = yes
  sanitizers    = 
  debug enabled = no
  gprof enabled = no
  werror        = no

  target os     = 
  build os      = 

  CC            = gcc
  CFLAGS        = -g -O2
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = g++ -std=c++11
  CXXFLAGS      =   -Wstack-protector -fstack-protector-all  -Wall -Wextra -Wformat -Wvla  -Wno-unused-parameter   -g -O2
  LDFLAGS       = -D_POSIX_PTHREAD_SEMANTICS -pthread  -Wl,-z,now -pie  
  ARFLAGS       = cr


/opt/local/src/bitcoin]# make
Making all in src
make[1]: Entering directory '/opt/local/src/bitcoin/src'
make[2]: Entering directory '/opt/local/src/bitcoin/src'
make[3]: Entering directory '/opt/local/src/bitcoin'
make[3]: Leaving directory '/opt/local/src/bitcoin'
  CXX      crypto/libbitcoinconsensus_la-aes.lo
  CXX      crypto/libbitcoinconsensus_la-chacha20.lo
  CXX      crypto/libbitcoinconsensus_la-hmac_sha256.lo
  CXX      crypto/libbitcoinconsensus_la-hmac_sha512.lo
  CXX      crypto/libbitcoinconsensus_la-ripemd160.lo
  CXX      crypto/libbitcoinconsensus_la-sha1.lo
  CXX      crypto/libbitcoinconsensus_la-sha256.lo
  CXX      crypto/libbitcoinconsensus_la-sha512.lo
  CXX      crypto/libbitcoinconsensus_la-sha256_sse4.lo
  CXX      libbitcoinconsensus_la-arith_uint256.lo
  CXX      consensus/libbitcoinconsensus_la-merkle.lo
In file included from ./script/script.h:11:0,
                 from ./primitives/transaction.h:11,
                 from ./consensus/merkle.h:11,
                 from consensus/merkle.cpp:5:
./serialize.h:193:39: error: redefinition of 'template<class Stream> void Serialize(Stream&, int8_t)'
 template<typename Stream> inline void Serialize(Stream& s, int8_t a  ) { ser_wr
                                       ^
./serialize.h:192:39: note: 'template<class Stream> void Serialize(Stream&, char)' previously declared here
 template<typename Stream> inline void Serialize(Stream& s, char a    ) { ser_wr
                                       ^
In file included from ./script/script.h:11:0,
                 from ./primitives/transaction.h:11,
                 from ./consensus/merkle.h:11,
                 from consensus/merkle.cpp:5:
./serialize.h:209:39: error: redefinition of 'template<class Stream> void Unserialize(Stream&, int8_t&)'
 template<typename Stream> inline void Unserialize(Stream& s, int8_t& a  ) { a =
                                       ^
./serialize.h:208:39: note: 'template<class Stream> void Unserialize(Stream&, char&)' previously declared here
 template<typename Stream> inline void Unserialize(Stream& s, char& a    ) { a =
                                       ^
Makefile:8138: recipe for target 'consensus/libbitcoinconsensus_la-merkle.lo' failed
make[2]: *** [consensus/libbitcoinconsensus_la-merkle.lo] Error 1
make[2]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:9824: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:756: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1


```

### PR DESCRIPTION
**Pull request by ken2812221**
Probably fixes #13576. I'm not able to test this. @stacepellegrino, can you test this?

### CODE REVIEW

**Comment by stacepellegrino** [comment]
Just testing changes now. Performing build to see if it fixes [https://github.com/bitcoin/bitcoin/issues/13576](url)

**Comment by stacepellegrino** [comment]
Did not fix the problem...

```
# make
Making all in src
make[1]: Entering directory '/opt/local/src/bitcoin/src'
make[2]: Entering directory '/opt/local/src/bitcoin/src'
make[3]: Entering directory '/opt/local/src/bitcoin'
make[3]: Leaving directory '/opt/local/src/bitcoin'
  CXX      crypto/libbitcoinconsensus_la-aes.lo
  CXX      crypto/libbitcoinconsensus_la-chacha20.lo
  CXX      crypto/libbitcoinconsensus_la-hmac_sha256.lo
  CXX      crypto/libbitcoinconsensus_la-hmac_sha512.lo
  CXX      crypto/libbitcoinconsensus_la-ripemd160.lo
  CXX      crypto/libbitcoinconsensus_la-sha1.lo
  CXX      crypto/libbitcoinconsensus_la-sha256.lo
  CXX      crypto/libbitcoinconsensus_la-sha512.lo
  CXX      crypto/libbitcoinconsensus_la-sha256_sse4.lo
  CXX      libbitcoinconsensus_la-arith_uint256.lo
  CXX      consensus/libbitcoinconsensus_la-merkle.lo
In file included from ./script/script.h:11:0,
                 from ./primitives/transaction.h:11,
                 from ./consensus/merkle.h:11,
                 from consensus/merkle.cpp:5:
./serialize.h:195:39: error: redefinition of 'template<class Stream> void Serialize(Stream&, int8_t)'
 template<typename Stream> inline void Serialize(Stream& s, int8_t a  ) { ser_wr
                                       ^
./serialize.h:193:39: note: 'template<class Stream> void Serialize(Stream&, char)' previously declared here
 template<typename Stream> inline void Serialize(Stream& s, char a    ) { ser_wr
                                       ^
In file included from ./script/script.h:11:0,
                 from ./primitives/transaction.h:11,
                 from ./consensus/merkle.h:11,
                 from consensus/merkle.cpp:5:
./serialize.h:213:39: error: redefinition of 'template<class Stream> void Unserialize(Stream&, int8_t&)'
 template<typename Stream> inline void Unserialize(Stream& s, int8_t& a  ) { a =
                                       ^
./serialize.h:211:39: note: 'template<class Stream> void Unserialize(Stream&, char&)' previously declared here
 template<typename Stream> inline void Unserialize(Stream& s, char& a    ) { a =
                                       ^
Makefile:8138: recipe for target 'consensus/libbitcoinconsensus_la-merkle.lo' failed
make[2]: *** [consensus/libbitcoinconsensus_la-merkle.lo] Error 1
make[2]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:9824: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:757: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

**Comment by stacepellegrino** [comment]
The configure script generates the following as part of its output...

`checking for if type char equals int8_t... no
`

**Comment by ken2812221** [comment]
Can you provide your config.log?

**Comment by ken2812221** [comment]
I forgot to specify the error message for static_assert, would u mind to test this again?

**Comment by stacepellegrino** [comment]
That works! Please update the main branch of source with this change.

However, my build failed with...

  CXX      compat/libbitcoin_util_a-glibc_sanity.o
In file included from compat/glibc_sanity.cpp:12:0:
compat/glibc_sanity.cpp: In function 'bool {anonymous}::sanity_test_fdelt()':
compat/glibc_sanity.cpp:53:5: error: 'memset' was not declared in this scope
     FD_ZERO(&fds);
     ^
Makefile:6248: recipe for target 'compat/libbitcoin_util_a-glibc_sanity.o' failed
make[2]: *** [compat/libbitcoin_util_a-glibc_sanity.o] Error 1
make[2]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:9824: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/opt/local/src/bitcoin/src'
Makefile:757: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1

Shall I open this as a new issue?

**Comment by stacepellegrino** [comment]
Checking for type char equals int8_t fixed.

Opened a new build issue here... [https://github.com/bitcoin/bitcoin/issues/13581](url)

File: configure.ac [hunk 1]
```diff
@@ -808,6 +808,14 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <sys/types.h>
  [ AC_MSG_RESULT(no)]
 )
 
+AC_MSG_CHECKING(for if type char equals int8_t)
+AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <stdint.h>
+  #include <type_traits>]],
+ [[ static_assert(std::is_same<int8_t, char>::value, ""); ]])],
```
Comment (Empact) [inline configure.ac]: nit: may as well provide a message for the static_assert?

**Comment by Empact** [comment]
utACK 49d1f4c

**Review by jb55 — APPROVED** [review]
utACK 49d1f4cdde6d3289cb8c18ad35fc739371e25388

**Comment by laanwj** [comment]
Can someone please explain why this is needed? I've been compiling bitcoin core on platforms that have char=uint8_t (such as ARM) as well as char=int8_t (such as x86) for years without needing any special define logic for this. What is different on this platform?

**Comment by luke-jr** [comment]
Apparently ARM and x86 have `int8_t` = `signed char`, whereas SunOS has `int8_t` = `char`?

**Comment by laanwj** [comment]
Thanks @luke-jr, I understand then.
utACK 49d1f4cdde6d3289cb8c18ad35fc739371e25388

File: src/serialize.h [hunk 2]
```diff
@@ -189,7 +189,9 @@ template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
         SerializationOp(s, CSerActionUnserialize());                  \
     }
 
+#ifndef CHAR_EQUALS_INT8
 template<typename Stream> inline void Serialize(Stream& s, char a    ) { ser_writedata8(s, a); } // TODO Get rid of bare char
```
Comment (laanwj) [inline src/serialize.h]: I like the TODO here - not sure its still the goal, certainly not in this PR, would have to carefully consider the consequences for serialization.

**Pull request merged**